Make project file clean.

frg2089 opened the pull request:
Visual Sutdio supports writing this in the project file.
However, it does not support easy switching between "`engine`" and "`game`" properties.

And I rewrite the build script in PowerShell Core.

Build State
|:-:|
[![build client](https://github.com/frg2089/xna-cncnet-client/actions/workflows/build.yml/badge.svg)](https://github.com/frg2089/xna-cncnet-client/actions/workflows/build.yml)

See [CnCNet/xna-cncnet-client#341](https://github.com/CnCNet/xna-cncnet-client/pull/341#discussion_r900169895)

ref CnCNet/xna-cncnet-client#341

Rans4ckeR commented:
@frg2089 I like the idea of the clean project files and the PS script.
I do miss however the existing solution configurations and platforms. Can we keep the original configurations in the solution file and in the Build.props? Otherwise development, debugging, building in Visual Studio is not possible for specific games and engines. Can you do a mapping between the existing solution configurations and your msbuild values? e.g. AresDXDebug to Game=Ares, Engine=DX, Configuration=Debug.

The build script is missing a few things. Maybe you can do a directory compare of the output with the existing BuildAll script to match the exact file and directory listings? Also check the Architecture of the startup assemblies. e.g. .NET 4.8 clientxna.exe and .NET 6.0 clientxna.dll have to be x86.

frg2089 committed:
why inside namespace?

C# Standard is outside namespace.

frg2089 committed:
update msbuild project

frg2089 committed:
use powershell script

frg2089 committed:
make Visual Studio Compatible

frg2089 committed:
add XNA build

frg2089 committed:
update build script

frg2089 commented:
> Can you do a mapping between the existing solution configurations and your msbuild values? e.g. AresDXDebug to Game=Ares, Engine=DX, Configuration=Debug.

Now, we can use `Platforms` property to switch `Game` and `Engine`.

such as
|Platform|Game|Engine|
|:-:|:-:|:-:|
AresDX|Ares|DX
TSGL|TS|GL
YRXNA|YR|XNA
...|...|...


> Also check the Architecture of the startup assemblies. e.g. .NET 4.8 clientxna.exe and .NET 6.0 clientxna.dll have to be x86.

I use the `PlatformTarget` property to make XNA an x86 architecture.
And then, we can build XNA use the build script.

And you may use `git rebase` command to make git commit clean.
Too many merge commit!

- 8d46715a748659597eea06107b1d34f8b831d536
- 77990d372a702b8c401e8ad632b57b76da2f556c

<details>
<summary> PowerShell Script ! </summary>

## Generate Platforms

```powershell
(@('Ares','TS','YR')|%{
  $_1 = $_
  @('DX','GL','XNA')|%{"$_1$_"}
})-join';'
```

## Generate SolutionConfigurationPlatforms Section

```powershell
@('Ares','TS','YR')|%{
  $_1 = $_
  @('DX','GL','XNA')|%{
    "Debug|$_1$_ = Debug|$_1$_"
    "Release|$_1$_ = Release|$_1$_"
  }
}|sort
```

## Generate ProjectConfigurationPlatforms Section

```powershell
@('Ares','TS','YR')|%{
  $_1 = $_
  @('DX','GL','XNA')|%{
    $_2 = $_
    @(
      '{97458C1E-2E6C-4C5C-93C7-16A6712802E9}'
      '{DAA331A7-FDAC-44FD-98B5-2DC5025F48C3}'
      '{4C26AD04-FF2E-4465-9F40-DEB9DDC2C1D6}'
      '{D517317D-8EA8-4225-AF9E-5385D8A7F047}'
      '{2CD114E3-C30F-4A95-9B4A-455C2C8F82E5}'
    )|%{
      "$_.Debug|$_1$_2.ActiveCfg = Debug|$_1$_2"
      "$_.Debug|$_1$_2.Build.0 = Debug|$_1$_2"
      "$_.Release|$_1$_2.ActiveCfg = Release|$_1$_2"
      "$_.Release|$_1$_2.Build.0 = Release|$_1$_2"
    }
  }
}|sort
```

</details>

Metadorius commented on build/Framework.props:
```diff
@@ -0,0 +1,38 @@
+  </PropertyGroup>
+
+  <Target Name="CheckEngineAndGame" BeforeTargets="CoreCompile">
+    <Error Text="Engine are not set! Please retry with /p:Engine={Engine}. Supported Engines is $(Engines)." Condition=" '$(Engine)' == '' " />
+    <Error Text="Engine '$(Engine)' are not support. Supported Engines is $(Engines)." Condition=" '!$(Engines.Contains($(Engine)))' " />
+
+    <Error Text="Game are not set! Please retry with /p:Game={Game}. Supported Games is $(Games)." Condition=" '$(Game)' == '' " />
+    <Error Text="Game '$(Game)' are not support. Supported Games is $(Games)." Condition=" '!$(Games.Contains($(Game)))' " />
```
```suggestion
    <Error Text="Engine is not set! Please retry with /p:Engine={Engine}. Supported Engines are $(Engines)." Condition=" '$(Engine)' == '' " />
    <Error Text="Engine '$(Engine)' is not supported. Supported Engines are $(Engines)." Condition=" '!$(Engines.Contains($(Engine)))' " />

    <Error Text="Game is not set! Please retry with /p:Game={Game}. Supported Games are $(Games)." Condition=" '$(Game)' == '' " />
    <Error Text="Game '$(Game)' is not supported. Supported Games are $(Games)." Condition=" '!$(Games.Contains($(Game)))' " />
```

Metadorius commented:
I like the idea of cross-platform build scripts but please add scripts to build only Ares, only TS and only YR binaries for convenience.

Rans4ckeR commented on DXClient.sln:
```diff
@@ -16,714 +16,234 @@ EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{0F60E4A3-55C8-4C3F-BFC0-27C597433E8A}"
 	ProjectSection(SolutionItems) = preProject
 		.editorconfig = .editorconfig
+		Directory.Build.props = Directory.Build.props
+		build\Framework.props = build\Framework.props
```
Can you make it work by only making these changes to the solution file? All other changes to the solution file I would like to revert. Without that specific solution configuration it is not possible to debug for example the code for the combination TS+XNA+Debug+x86 or DX+GL+Release+x64 etc. in VS.

Rans4ckeR replied:
There are still some changes that should be reverted. When you open Configuration Manager in VS and select any XNA configuration as active configuration in combination with any active platform other than x86 the Build checkbox should be disabled.
It should be safe to just copy paste the contents of the old file and add the above 2 lines again I believe.

Rans4ckeR commented on ClientCore/ClientCore.csproj:
```diff
@@ -14,58 +11,16 @@
-    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
-    <Platforms>AnyCPU;x64;ARM64;x86</Platforms>
     <GenerateResourceUsePreserializedResources>true</GenerateResourceUsePreserializedResources>
-    <Configurations>AresDXDebug;AresGLDebug;AresXNADebug;AresDXRelease;AresGLRelease;AresXNARelease;TSDXDebug;TSGLDebug;TSXNADebug;TSDXRelease;TSGLRelease;TSXNARelease;YRDXDebug;YRGLDebug;YRXNADebug;YRDXRelease;YRGLRelease;YRXNARelease</Configurations>
```
See comment on solution file, we need this specific configuration for all projects. It can be moved to a props file of course.

Rans4ckeR commented:
> I like the idea of cross-platform build scripts but please add scripts to build only Ares, only TS and only YR binaries for convenience.

Agreed, I would say at the minimum keep the granularity of the existing scripts:
![image](https://user-images.githubusercontent.com/25006126/174452527-83cd4a69-1964-4ae0-82d0-bc035fa42cab.png)

Rans4ckeR commented on build/VSCompatibleLayer.props:
```diff
@@ -0,0 +1,16 @@
+<Project>
+  <PropertyGroup>
+    <Platforms>AresDX;AresGL;AresXNA;TSDX;TSGL;TSXNA;YRDX;YRGL;YRXNA</Platforms>
```
See comment on solution file. We should keep the existing platform values AnyCPU;x64;ARM64;x86.

Also from [Common MSBuild project properties](https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-properties?view=vs-2022):
`Platform:
The operating system you are building for. Examples for .NET Framework builds are "Any CPU", "x86", and "x64".'`
The VS Configuration Manager for example does not allow you to use other values than the predefined ones:
![image](https://user-images.githubusercontent.com/25006126/174454452-4e5ff1d5-751c-454d-a470-8162c8299ba8.png)

Rans4ckeR commented:
I'm getting this error multiple times when running buildAll.ps1:

![image](https://user-images.githubusercontent.com/25006126/174453248-61a8d955-c308-456a-80a6-012381fdce75.png)

frg2089 committed:
Update build/Framework.props

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

frg2089 committed:
refactor build script

frg2089 committed:
refactor Visual Studio Compatible

frg2089 committed:
fix error

frg2089 committed:
fix ci

frg2089 committed:
Use different assembly names for different engines

frg2089 committed:
fixed build script

`. ` import cannot precede the param block.

However, the string can be implicitly converted to an enumerated value.

frg2089 committed:
fixed an error

frg2089 commented:
> I like the idea of cross-platform build scripts but please add scripts to build only Ares, only TS and only YR binaries for convenience.

> I'm getting this error multiple times when running buildAll.ps1:
> 
> ![image](https://user-images.githubusercontent.com/25006126/174453248-61a8d955-c308-456a-80a6-012381fdce75.png)

I refactored the build script and it works fine now.


<details>
<summary> PowerShell Script ! </summary>

## Generate SolutionConfigurationPlatforms Section

```powershell
@('Ares','TS','YR')|%{
  $_1 = $_
  @('DX','GL','XNA')|%{
    $_2 = $_
    @('Debug','Release')|%{
      $_3 = $_
      @('Any CPU','x86', 'x64', 'ARM64')|%{
        "$_1$_2$_3|$_ = $_1$_2$_3|$_"
      }
    }
  }
}|sort|clip
```

## Generate ProjectConfigurationPlatforms Section

```powershell
@('Ares','TS','YR')|%{
  $_1 = $_
  @('DX','GL','XNA')|%{
    $_2 = $_
    @('Debug','Release')|%{
      $_3 = $_
      @('Any CPU','x86', 'x64', 'ARM64')|%{
        $_4 = $_
        @(
          '{97458C1E-2E6C-4C5C-93C7-16A6712802E9}'
          '{DAA331A7-FDAC-44FD-98B5-2DC5025F48C3}'
          '{4C26AD04-FF2E-4465-9F40-DEB9DDC2C1D6}'
          '{D517317D-8EA8-4225-AF9E-5385D8A7F047}'
          '{2CD114E3-C30F-4A95-9B4A-455C2C8F82E5}'
        )|%{
          "$_.$_1$_2$_3|$_4.ActiveCfg = $_1$_2$_3|$_4"
          "$_.$_1$_2$_3|$_4.Build.0 = $_1$_2$_3|$_4"
        }
      }
    }
  }
}|sort|clip
```
然而PowerShell，它棒极了！
用PowerShell生成解决方案节避免了我枯燥的手改解决方案！
</details>

Rans4ckeR commented on BuildScripts/lib/BuildCore.ps1:
```diff
@@ -0,0 +1,70 @@
+    $Private:ClientSuffix = Get-Suffix $Engine
+
+    $Private:DotnetArgs = @(
+      "build"
```
Please use dotnet publish instead of dotnet build. dotnet publish is the only official way to make a 'production ready' release.
![image](https://user-images.githubusercontent.com/25006126/174470995-f7156763-b714-4dbd-86bf-579e1316bbe0.png)

frg2089 committed:
use dotnet publish

frg2089 committed:
disable build XNA when current platform are not x86

Rans4ckeR commented:
> The build script is missing a few things. Maybe you can do a directory compare of the output with the existing BuildAll script to match the exact file and directory listings?

- The common files are copied to \Binaries and \Binaries\<Get-PlatformName>. They should only be present in \Binaries.
- The .NET 6 clientdx.* , clientogl.* and clientxna.* files should all be copied to \Binaries\<Get-PlatformName>, except for *.exe. The 3 .exe files should be the only files present in the root directory.
- The .NET 4.8 \Binaries\OpenGL directory should not contain SDL2.dll and soft_oal.dll so that the correct x86 or x64 version is loaded at runtime from the \x86 or \x64 directory.
- The .NET 6 client*.exe AppHost files need to be patched so they point to the correct dll filename and directory path e.g. \Binaries\Windows\clientdx.dll after moving those. From the current script:
   `..\AdditionalFiles\AppHostPatcher\AppHostPatcher.exe ..\Compiled\%1\net60\Resources\clientdx.exe DXMainClient.dll Binaries\Windows\clientdx.dll`
- Can we set the output path to e.g. \Compiled\YR\net48\Resources instead of \Compiled\YR\net48? I find it easier so that after the build we can copy all content files to e.g. \Compiled\YR\net48\ and be ready to run instead of having to manually create the correct directory structure.

frg2089 committed:
MSBuild AppHostPatcher

frg2089 commented:
How should we do?

.Net 6:
publish to `Compiled\$Game\net6.0\Binaries\$Engine` (MSBuild will be patch the `exe` file.)
Move Common Libraries to `Compiled\$Game\net6.0\Binaries`
Move `*.exe` to `Compiled\$Game\net6.0`

.Net Framework 4.8:
publish to `Compiled\$Game\net48\Binaries\$Engine`
Move Common Libraries to `Compiled\$Game\net48\Binaries`
Move `client*.*` to `Compiled\$Game\net48`
When `$Engine` is `OpenGL` remove `SDL2.dll` and `soft_oal.dll`

right?

Rans4ckeR commented:
> .Net 6: 
> publish to `Compiled\$Game\net6.0\Binaries\$Engine` (MSBuild will be patch the `exe` file.) 

The current script uses: `Compiled\$Game\net6.0\Resources\Binaries\$Engine`

> Move Common Libraries to `Compiled\$Game\net6.0\Binaries` 

The current script uses: `Compiled\$Game\net6.0\Resources\Binaries`

> Move `*.exe` to `Compiled\$Game\net6.0`

The current script uses: `Compiled\$Game\net6.0\Resources`

> .Net Framework 4.8:
> publish to `Compiled\$Game\net48\Binaries\$Engine`

The current script uses: `Compiled\$Game\net48\Resources\Binaries\$Engine`

> Move Common Libraries to `Compiled\$Game\net48\Binaries` 

The current script uses: `Compiled\$Game\net48\Resources\Binaries`

> Move `client*.*` to `Compiled\$Game\net48` 

The current script uses: `Compiled\$Game\net48\Resources`

> When `$Engine` is `OpenGL` remove `SDL2.dll` and `soft_oal.dll`

yes

frg2089 committed:
update targets

frg2089 committed:
refactor common file script

frg2089 committed:
refactor build script

frg2089 committed:
update

Rans4ckeR commented:
I assume you're building on linux? Can you also try on Windows?
Only the Compiled directory is created, not the subdirectories.

It currently doesn't show what went wrong:
> Exception: C:\Users\RansAckeR\source\repos\frg2089\xna-cncnet-client\BuildScripts\lib\BuildCore.ps1:60
Line |
  60 |        throw "Failed to build project"
     |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Failed to build project

And it continues even if an error occurs.

If I add:

```PS
$ErrorActionPreference = "Stop"

trap
{
	$Error
	$_ | Format-List -Force | Out-String
	$_.Exception | Format-List -Force | Out-String
	"Message=Error: " + $Error
	Start-Sleep -s 10
    Write-Error $_
    Exit 1
}
```
and change
```PS
if ($LASTEXITCODE -ne 0) {
```
to
```PS
if ($LASTEXITCODE -gt 1) {
```

It shows:
> Write-Error: Cannot find path 'C:\Users\RansAckeR\source\repos\frg2089\xna-cncnet-client\Compiled\Ares\net6.0\Resources\Binaries\Windows\clientdx.exe' because it does not exist.

It seems the default exit code for dotnet is 1.

frg2089 commented:
> I assume you're building on linux? Can you also try on Windows?

My computer's operating system is Windows.
I have not tried to build it in Linux.

It works fine on my computer and on Github Actions.

You can see [this ci log](https://github.com/frg2089/xna-cncnet-client/runs/6977063475?check_suite_focus=true).
I don't have mono configured for Ubuntu and Mac so it won't work properly.
But it works fine on Windows.

I disabled all code style warnings in commit 6a875bd to make the output clean.

Rans4ckeR commented:
The script errors are fixed for me now.

Can we however inverse the mechanism for common files? I would work with a list of files per Engine that need to be copied to the $Engine directory, not a list of common files that need to be copied to \Binaries. This is also what the current build scripts do.
The reason for this is that the Engine file list will rarely change while the common file list is very dynamic towards future changes. Otherwise it will be easy to miss new files and the build scripts will need to be reviewed every time.

Can we work with 2 .editorconfig files? One for Debug enabling all warnings and one for Release that has code style warnings disabled?

_FYI
These files should be shared (Compiled\\$Game\\$TargetFramework\Resources\Binaries)._
> Compiled\\$Game\\$TargetFramework\Resources\Binaries\\$Engine\Localization.pdb
> Compiled\\$Game\\$TargetFramework\Resources\Binaries\\$Engine\Localization.dll

_And these seem to be missing for .NET 4.8 (Compiled\\$Game\net48\Resources\Binaries):_
> System.Buffers.dll
> System.Memory.dll
> System.Numerics.Vectors.dll
> System.Resources.Extensions.dll
> System.Runtime.CompilerServices.Unsafe.dll

frg2089 committed:
refactor build script

frg2089 commented:
> Can we work with 2 .editorconfig files? One for Debug enabling all warnings and one for Release that has code style warnings disabled?

I don't think it's a good idea that the project should use only one `.editorconfig` file.

Instead of using two different `.editorconfig` files, the project should be refactored to make its code style the same as `.editorconfig` to eliminate warnings.

~~There is a secondary reason why I cann't do it.~~


If you think this PR meets the requirements for merging, please remind me to delete these submissions to keep the submission record clean.
- 6a875bd
- ac582ee
- 8f80f54

frg2089 committed:
fix encoding error

see https://github.com/CnCNet/xna-cncnet-client/pull/341#issuecomment-1162540275

Rans4ckeR commented on BuildScripts/lib/SpecialDXList.txt:
```diff
@@ -0,0 +1,28 @@
+ClientGUI.pdb
+DTAConfig.dll
+DTAConfig.pdb
+Localization.dll
+Localization.pdb
```
```suggestion
```

Rans4ckeR commented on BuildScripts/lib/SpecialGLList.txt:
```diff
@@ -0,0 +1,21 @@
+clientogl.runtimeconfig.json
+DTAConfig.dll
+DTAConfig.pdb
+Localization.dll
+Localization.pdb
```
```suggestion
```

Rans4ckeR commented on BuildScripts/lib/SpecialXNAList.txt:
```diff
@@ -0,0 +1,25 @@
+clientxna.runtimeconfig.json
+DTAConfig.dll
+DTAConfig.pdb
+Localization.dll
+Localization.pdb
```
```suggestion
```

frg2089 committed:
Update build/AfterPublish.targets

Co-authored-by: Rans4ckeR <ransacker@outlook.com>

Rans4ckeR commented on build/AfterPublish.targets:
```diff
@@ -0,0 +1,18 @@
+    <Exec ConsoleToMsBuild="True" Command="&quot;$(MSBuildThisFileDirectory)..\AdditionalFiles\AppHostPatcher\AppHostPatcher.exe&quot; &quot;$(PublishDir)$(AssemblyName).exe&quot; &quot;$(AssemblyName).dll&quot; &quot;Binaries\$(__DirectoryName)\$(AssemblyName).dll&quot;" />
+  </Target>
+
+  <Target Name="RemoveSDL" AfterTargets="Publish" Condition="$(TargetFrameworkIdentifier)' == '.NETFramework'">
```
```suggestion
  <Target Name="RemoveSDL" AfterTargets="Publish" Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
```

Rans4ckeR commented on ClientCore/PlatformShim/EncodingExt.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    static EncodingExt()
+    {
+#if NETCOREAPP3_0_OR_GREATER
```
```suggestion
#if !NETFRAMEWORK
```

Rans4ckeR commented on ClientCore/PlatformShim/EncodingExt.cs:
```diff
@@ -0,0 +1,17 @@
+using System;
```
```suggestion
```

Rans4ckeR reviewed (changes requested):
The syntax error prevents the build from working atm.
The Localization.dll is the same for all engines and can be shared.

frg2089 committed:
Update ClientCore/PlatformShim/EncodingExt.cs

Co-authored-by: Rans4ckeR <ransacker@outlook.com>

frg2089 committed:
Update BuildScripts

Co-authored-by: Rans4ckeR <ransacker@outlook.com>

Rans4ckeR commented:
One last request for debugging purposes. Can you make it so the build outputs of different games don't overwrite each other?
For example when I build YRDXDebug, the output is going here:
> \DXMainClient\bin\Debug\DX\net6.0-windows10.0.22000.0
> \DXMainClient\bin\Debug\DX\net48

I then build AresDXDebug and all files are overwritten in the same directory.
So when I switch back to YRDXDebug I'm currently forced to do a rebuild.

The previous build used these paths:
> DXMainClient\bin\YRDXDebug\net6.0-windows10.0.22000.0
> DXMainClient\bin\YRDXDebug\net48

So perhaps like this?
> \DXMainClient\bin\Debug\\$Game\DX\net6.0-windows10.0.22000.0
> \DXMainClient\bin\Debug\\$Game\DX\net48

frg2089 committed:
add $Game in output path

frg2089 committed:
fix CA2200

Rans4ckeR commented:
@frg2089 

> If you think this PR meets the requirements for merging, please remind me to delete these submissions to keep the submission record clean.
> 
> * [6a875bd](https://github.com/Rans4ckeR/xna-cncnet-client/commit/6a875bd6ab0233aedbd46540367336a79c20ec2c)
> * [ac582ee](https://github.com/Rans4ckeR/xna-cncnet-client/commit/ac582ee1283051cb58094cef0618d806cdf99718)
> * [8f80f54](https://github.com/Rans4ckeR/xna-cncnet-client/commit/8f80f549eff6b53d1f62a88fa1dd845c81bf8bf2)
